### ISSUE TITLE
PDAL 2.6 sometimes drops CRS during conversion

### ISSUE THREAD

**Issue by olsen232** [PDAL/PDAL#4191]
**Describe the bug**
(Unreleased) PDAL 2.6 has new CRS extraction behaviour that is more capable but perhaps less robust to strange inputs. If you use PDAL 2.6 to convert `autzen.las` to COPC, it drops the attached CRS completely.

Steps to reproduce:
Find `autzen.las` from the LibLAS project: https://github.com/libLAS/libLAS/blob/master/test/data/autzen.las
Put it in a directory with a pipeline to convert it to COPC. Here's mine:
```
[
    {
        "type": "readers.las",
        "filename": "autzen.las"
    },
    {
        "type": "writers.copc",
        "filename": "autzen.copc.laz",
        "forward": "all"
    }
]
```
Run the pipeline using PDAL 2.5.6 (or any PDAL built from the 2.5 maintenance branch). Then run `pdal info --metadata` on the result. Observe that there is lots of CRS output (horizontal CRS, compound CRS, etc etc). However, note that this CRS is not the same as the original CRS from autzen.las, but instead is loaded from the PROJ DB - that is not this bug though, that is just a limitation in 2.5.x versions of PDAL.

Run the pipeline using PDAL 2.6 (that is, a PDAL built from master). Then run `pdal info --metadata` on the result. Observe that there is no CRS output at all.

Running the unix command `strings` is also a way of detecting whether a LAS / LAZ file has an embedded CRS - at least if it has a WKT one in the VLRs. The original file and the 2.5.6 conversion both have embedded WKT that shows up in strings, the 2.6 conversion does not.

**Expected behavior**
Some form of the CRS should be preserved and copied in some form to the converted output, even if the original CRS is a bit weird. (See last section below).

**Culprit:**
As far as I can ascertain, the behaviour changed in this commit:
https://github.com/PDAL/PDAL/commit/040cd8c7046f4354a0476ae0e40e0ba0fa6beb12
My guess is that previously, PDAL didn't really use the embedded WKT - only the Geotiff VLRs - and so it didn't really have to deal with WKT was unusual, since all the WKT it ever actually used was loaded from the PROJ DB.

**System/installation information:**
Please provide information on your PDAL version (`pdal --version`) and system (e.g., `uname -a` or `ver`).

```
$ uname -a
Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64 x86_64
```

```
$ pdal --version
pdal 2.6.0 (git-version: b902a3)
```

**Slightly strange WKT of autzen.las**

autzen.las has the following WKT for EPSG:2994, which differs a bit from what you'd find in PROJ DB. I'm not sure if it conforms to the WKT spec or not. It may be that something about it is tripping up PDAL. One possible resolution is to decide that this WKT and therefore the file autzen.las is not conformant and not supported - I wouldn't know - all I know is, it is part of the libLAS test data.

```
PROJCS["NAD83(HARN) / Oregon Lambert (ft)",
    GEOGCS["NAD83(HARN)",
        DATUM["NAD83_High_Accuracy_Regional_Network",
            SPHEROID["GRS 1980", 6378137, 298.257222101,
                AUTHORITY["EPSG", "7019"]],
            TOWGS84[0, 0, 0, 0, 0, 0, 0],
            AUTHORITY["EPSG", "6152"]],
        PRIMEM["Greenwich", 0,
            AUTHORITY["EPSG", "8901"]],
        UNIT["degree", 0.0174532925199433,
            AUTHORITY["EPSG", "9122"]],
        AUTHORITY["EPSG", "4152"]],
    UNIT["foot", 0.3048,
        AUTHORITY["EPSG", "9002"]],
    PROJECTION["Lambert_Conformal_Conic_2SP"],
    PARAMETER["standard_parallel_1", 43],
    PARAMETER["standard_parallel_2", 45.5],
    PARAMETER["latitude_of_origin", 41.75],
    PARAMETER["central_meridian", -120.5],
    PARAMETER["false_easting", 1312335.958],
    PARAMETER["false_northing", 0],
    AUTHORITY["EPSG", "2994"],
    AXIS["X", EAST],
    AXIS["Y", NORTH]]
```

**Comment by hobu** [comment]
@jjimenezshaw can you take a look at this?

**Comment by hernando** [comment]
The string is out of order.

According to the WKT 2 spec for a projected CRS (https://docs.ogc.org/is/18-010r7/18-010r7.html#72) after the base geodetic definition it goes the map projection. In other words, `UNIT` shouldn't be there. `AUTHORITY` is also out of place.
That said, the WKT string above is version 1, not 2, but I suppose the same rules apply.

A conforming WKT would be:
```
PROJCS["NAD83(HARN) / Oregon Lambert (ft)",
    GEOGCS["NAD83(HARN)",
        DATUM["NAD83_High_Accuracy_Regional_Network",
            SPHEROID["GRS 1980", 6378137, 298.257222101,
                AUTHORITY["EPSG", "7019"]],
            TOWGS84[0, 0, 0, 0, 0, 0, 0],
            AUTHORITY["EPSG", "6152"]],
        PRIMEM["Greenwich", 0,
            AUTHORITY["EPSG", "8901"]],
        UNIT["degree", 0.0174532925199433,
            AUTHORITY["EPSG", "9122"]],
        AUTHORITY["EPSG", "4152"]],
    PROJECTION["Lambert_Conformal_Conic_2SP"],
    PARAMETER["standard_parallel_1", 43],
    PARAMETER["standard_parallel_2", 45.5],
    PARAMETER["latitude_of_origin", 41.75],
    PARAMETER["central_meridian", -120.5],
    PARAMETER["false_easting", 1312335.958],
    PARAMETER["false_northing", 0],
    UNIT["foot", 0.3048,
        AUTHORITY["EPSG", "9002"]],
    AXIS["X", EAST],
    AXIS["Y", NORTH],
    AUTHORITY["EPSG", "2994"]]
```
At least `projinfo` doesn't complain with that one.

**Comment by olsen232** [comment]
Fair enough - what's the preferred PDAL behaviour here when eg converting the tile to a different format?
a) Drop nonconformant CRS, final result has no CRS (current behaviour)
b) Preserve nonconformant CRS verbatim in the final result
c) Try to repair / normalise the CRS during conversion (ie, in this case, either find EPSG:2994 in the WKT and use that code to replace the CRS with some canonical WKT from the PROJ DB - or, fallback to the lower-priority CRS source if the CRS from the first source is broken, which would give much the same result)

In addition to any of the above, could be good to log a warning

**Comment by jjimenezshaw** [comment]
I will try to find out why it behaves differently between 2.5 and master. I'm busy now, so probably I cannot until next week.

**Comment by abellgithub** [comment]
I will take care of this shortly.

**Comment by jjimenezshaw** [comment]
After some debuging, I think I got some clues (but I do not know how to solve the problem).

The copc file is not writing the metadata because the validation of the crs is not succeeding
```c++
void Output::addWktVlr()
{
    if (!b.srs.valid())
        return;
```

Why? because while reading the LAS file, in `Srs::init` it is using the wkt instead of the geotiff vlrs. That is because the input parameter `useWkt` was replaced by `srsConsumePreference`. In this case `useWkt` was false, so it was using the geotiff tags, not the wrongly defined wkt.

What is the proper solution? Passing both `useWkt` and `srsConsumePreference` to `Srs::init`? (I am not sure to understand the use-case of `useWkt`) Anything else?

BTW, if I am not wrong, reading with the option 
```
    "srs_consume_preference": "geotiff, wkt1"
```
should produce the expected result (copc file with srs info).

**Comment by abellgithub** [comment]
I will fix this tomorrow.

**Comment by olsen232** [comment]
Yes - I can confirm that putting "geotiff" ahead of "wkt1" in the srs_consume_preference fixes this particular issue (that is, causes a valid WKT to be read from autzen.las and written to autzen.copc.laz). Of course, improving how PDAL copes with malformed CRS will fix this issue more generally than switching the preference order of CRS extraction to prefer less expressive sources that are less likely to contain malformed CRS (I don't actually know whether the geotiff header can ever contain malformed CRS or not).

**Comment by jjimenezshaw** [comment]
About the "validity" of a CRS (used in `bool SpatialReference::valid() const`), GDAL is not "completely compliant" with WKT2. The return code can have these values:
``` 
return OGRERR_NONE if all is fine, OGRERR_CORRUPT_DATA if the SRS is
not well formed, and OGRERR_UNSUPPORTED_SRS if the SRS is well formed,
but contains non-standard PROJECTION[] values.
```
So some WKT2 are tagged as `OGRERR_UNSUPPORTED_SRS`, that is not `OGRERR_NONE`
(I do not mean that this is a bug, but something to keep in mind)

**Comment by hobu** [comment]
We are going to implement the following changes:

* Change the option from `srs_consume_preference` to `srs_vlr_order` because it's shorter more explicit.
* If `nosrs` is specified, don't read any CRS information.
* If `srs_vlr_order` is set, go through each of the specified possible VLRs, finding the first one that exists in the file and process that CRS.
* If the `LASFWkt2recordId` VLR is found, process that.
* If the `PdalProjJsonRecordId` VLR is found, process that.
* Process the WKT1 or geotiff VLR as determined by the global encoding flag (legacy behavior).

**Comment by hobu** [comment]
Addressed in #4205. Please reopen if we still have issues with control of this.
